compiler/qsc_eval/src/state.rs

```diff
@@ -0,0 +1,517 @@
+    real_part: RealNumber, // Real part
+    imaginary_part: RealNumber, // Imaginary part
+}
+
+impl CartesianForm {
+    fn construct(real_part: RealNumber, imaginary_part: RealNumber) -> Self {
+        let sign = real_part.sign();
+        match sign {
+            0 => Self {
+                sign: imaginary_part.sign(),
+                real_part: RealNumber::Zero,
+                imaginary_part: imaginary_part.abs(),
+            },
+            1.. => Self {
+                sign,
+                real_part,
+                imaginary_part,
+            },
+            _ => Self {
+                sign,
+                real_part: real_part.abs(),
+                imaginary_part: imaginary_part.negate(),
+            },
+        }
+    }
+
+    fn recognize(re: f64, im: f64) -> Self {
+        Self::construct(RealNumber::recognize(re), RealNumber::recognize(im))
+    }
+}
+
+// Represents a non-zero complex number which can be in either Polar or CartesianForm
+#[derive(Debug)]
+enum ComplexNumber {
+    Polar(PolarForm),
+    Cartesian(CartesianForm),
+}
+
+impl ComplexNumber {
+    fn recognize(re: f64, im: f64) -> Self {
+        if let Some(exponent) = PolarForm::recognize(re, im) {
+            Self::Polar(exponent)
+        } else {
+            Self::Cartesian(CartesianForm::recognize(re, im))
+        }
+    }
+}
+
+// Represents one term of a quantum state which corresponds to one basis vector
+struct Term {
+    basis_vector: BigUint,
+    coordinate: ComplexNumber,
+}
+
+fn get_terms_for_state(state: &Vec<(BigUint, Complex64)>) -> Vec<Term> {
+    let mut result: Vec<Term> = Vec::with_capacity(state.len());
+    for (basis_vector, coefficient) in state {
+        result.push(Term {
+            basis_vector: basis_vector.clone(),
+            coordinate: ComplexNumber::recognize(coefficient.re, coefficient.im),
+        });
+    }
+    result
+}
+
+#[must_use]
+pub fn get_latex(state: &Vec<(BigUint, Complex64)>, qubit_count: usize) -> String {
+    if state.len() > 16 {
+        return String::new();
+    }
+
+    let terms: Vec<Term> = get_terms_for_state(state);
+    let mut bad_term_count: i64 = 0;
+    for term in &terms {
+        if let ComplexNumber::Cartesian(cartesian) = &term.coordinate {
+            if let RealNumber::Decimal(_) = &cartesian.real_part {
+                bad_term_count += 1;
+            }
+            if let RealNumber::Decimal(_) = &cartesian.imaginary_part {
+                bad_term_count += 1;
+            };
+            if bad_term_count > 2 {
+                return String::new();
+            }
+        }
+    }
+
+    let mut latex: String = String::with_capacity(200);
+    latex.push_str("$|\\psi\\rangle = ");
+    let mut is_first: bool = true;
+    for term in terms {
+        write_latex_for_term(&mut latex, &term, !is_first);
+        let basis_label = fmt_basis_state_label(&term.basis_vector, qubit_count);
+        write!(latex, "|{basis_label}\\rangle").unwrap();
```

Comment:
niptick: not sure if these warnings are already coming up for you, but we disallowed use of `.unwrap()` across the codebase. You should use `.expect("... should ...")` instead where the message clarifies what the assumption was.

Reply:
This one is interesting. Along with these write! macros I'm using '.push' and '.push_str' functions. They don't require any special treatment in case of failure. So, I'm not sure how this write! would be different. The only problem seems to be when the process runs out of allotted memory. In this case not sure if any error message would be even generated from "expect" clause.
That said I don't see any reason not to change ".unwrap" to ".expect" and to provide some short string there.